File: csharp/DesiArchitect.DesignPatterns/03_Observer/Before/OrderManager.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Observer.Before;

// One god method that handles every downstream action when status changes
public class OrderManager
{
    public void UpdateOrderStatus(Order order, string newStatus)
    {
        order.Status = newStatus;
        Console.WriteLine($"  [Before] Order {order.Id} status -> {newStatus}");

        // SMS notification - hardcoded inline (simulated slow gateway blocks everything below)
        Console.WriteLine($"  [Before] SMS gateway slow (5s)...");
        Thread.Sleep(5000);
        Console.WriteLine($"  [Before] Sending SMS to {order.Phone}: Your order is now {newStatus}");

        // Analytics tracking - crammed in the same method
        Console.WriteLine($"  [Before] Tracking analytics: Order {order.Id} -> {newStatus}");

        // Rider notification - yet another responsibility
        Console.WriteLine($"  [Before] Notifying rider about order {order.Id}");

        // Dashboard refresh inlined in the same method
        Console.WriteLine($"  [Before] Refreshing dashboard for order {order.Id}");

        // New notification channel requires editing this method again
        // SMS service is down? Analytics stops too - everything is coupled.
    }
}

File: typescript/src/03-observer/before/order-manager.ts
import type { Order } from '../../shared/models.js';

// Synchronous sleep - blocks the event loop, the Node analog of C#'s Thread.Sleep.
function sleepSync(ms: number): void {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
}

// One god method that handles every downstream action when status changes.
export class OrderManager {
  updateOrderStatus(order: Order, newStatus: string): void {
    order.status = newStatus;
    console.log(`  [Before] Order ${order.id} status -> ${newStatus}`);

    // SMS notification inline - a slow gateway blocks everything below it.
    console.log('  [Before] SMS gateway slow (5s)...');
    sleepSync(5000);
    console.log(`  [Before] Sending SMS to ${order.phone}: Your order is now ${newStatus}`);

    // Analytics crammed into the same method.
    console.log(`  [Before] Tracking analytics: Order ${order.id} -> ${newStatus}`);

    // Rider notification - yet another responsibility.
    console.log(`  [Before] Notifying rider about order ${order.id}`);

    // Dashboard refresh inlined in the same method
    console.log(`  [Before] Refreshing dashboard for order ${order.id}`);

    // New channel requires editing this method again. Slow SMS blocks analytics too.
  }
}
